### PR TITLE
Generate enum constant classes for Compute

### PR DESCRIPTION
**Pull request by jskeet**
As ever, review one commit at a time.

### CODE REVIEW

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Cloud.Compute.V1.EnumConstantGenerator
+{
+    /// <summary>
+    /// A container for an either enums or other nested enum containers.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:25]:
```suggestion
    /// A container for either enums or other nested enum containers.
```

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            writer.WriteLine($"{indentation}/// <summary>Constants for wire representations of the <see cref=\"global::{enumType.FullName.Replace("+", ".")}\"/> enum.</summary>");
+            writer.WriteLine($"{indentation}public static class {descriptor.ClrType.Name}");
+            writer.WriteLine($"{indentation}{{");
+            writer.WriteLine();
+            string moreIndentation = indentation + "    ";
+            var formatMethod = typeof(ComputeEnumHelpers).GetMethod("Format").MakeGenericMethod(enumType);
+            foreach (var field in enumType.GetFields(BindingFlags.Static | BindingFlags.Public))
+            {
+                var wireValue = formatMethod.Invoke(null, new object[] { field.GetValue(null) });
+                Console.WriteLine($"{moreIndentation}/// <summary>Wire representation of <see cref=\"global::{enumType.FullName.Replace("+", ".")}.{field.Name}\"/>.</summary>");
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:84]: Maybe add the documentation of the Enume field as well, either as part of the summary or on remarks.

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            writer.WriteLine();
+        }
+
+        private static void GenerateEnumConstants(TextWriter writer, EnumDescriptor descriptor, string indentation)
+        {
+            var enumType = descriptor.ClrType;
+            writer.WriteLine($"{indentation}/// <summary>Constants for wire representations of the <see cref=\"global::{enumType.FullName.Replace("+", ".")}\"/> enum.</summary>");
+            writer.WriteLine($"{indentation}public static class {descriptor.ClrType.Name}");
+            writer.WriteLine($"{indentation}{{");
+            writer.WriteLine();
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:78]: nit: I think I'd remove this one.

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            string moreIndentation = indentation + "    ";
+            var formatMethod = typeof(ComputeEnumHelpers).GetMethod("Format").MakeGenericMethod(enumType);
+            foreach (var field in enumType.GetFields(BindingFlags.Static | BindingFlags.Public))
+            {
+                var wireValue = formatMethod.Invoke(null, new object[] { field.GetValue(null) });
+                Console.WriteLine($"{moreIndentation}/// <summary>Wire representation of <see cref=\"global::{enumType.FullName.Replace("+", ".")}.{field.Name}\"/>.</summary>");
+                Console.WriteLine($"{moreIndentation}public const string {field.Name} = \"{wireValue}\";");
+                Console.WriteLine();
+            }
+            writer.WriteLine($"{indentation}}}");
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:88]: nit: and maybe
```suggestion
            writer.Write($"{indentation}}}");
```

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1/ComputeEnumConstants.g.cs [hunk 5]
```diff
@@ -0,0 +1,5204 @@
...
+    {
+
+        /// <summary>Container class for enums within the <see cref="global::Google.Cloud.Compute.V1.AccessConfig"/> message.</summary>
+        public static class AccessConfig
+        {
+
+            /// <summary>Constants for wire representations of the <see cref="global::Google.Cloud.Compute.V1.AccessConfig.Types.NetworkTier"/> enum.</summary>
+            public static class NetworkTier
+            {
+
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1/ComputeEnumConstants.g.cs:28]: I left two comments on the generator that I think will remove this blank line...

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1/ComputeEnumConstants.g.cs [hunk 6]
```diff
@@ -0,0 +1,5204 @@
...
+                /// <summary>Wire representation of <see cref="global::Google.Cloud.Compute.V1.AccessConfig.Types.NetworkTier.UndefinedNetworkTier"/>.</summary>
+                public const string UndefinedNetworkTier = "UNDEFINED_NETWORK_TIER";
+
+                /// <summary>Wire representation of <see cref="global::Google.Cloud.Compute.V1.AccessConfig.Types.NetworkTier.Premium"/>.</summary>
+                public const string Premium = "PREMIUM";
+
+                /// <summary>Wire representation of <see cref="global::Google.Cloud.Compute.V1.AccessConfig.Types.NetworkTier.Standard"/>.</summary>
+                public const string Standard = "STANDARD";
+
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1/ComputeEnumConstants.g.cs:37]: ... and this one. All of that is of course a nit.

**Review by amanda-tarafa — APPROVED** [review]
LGTM

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:25]: Done.

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:78]: Have changed it to only add a blank line where necessary, both within containers and enums themselves.

File: apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Compute.V1/Google.Cloud.Compute.V1.EnumConstantGenerator/EnumContainer.cs:84]: That would be lovely, but significantly more work. I suggest we go with this for now, and consider improving later. I'll add a TODO.

**Comment by jskeet** [comment]
(The changes described are still in progress, but should be ready soon. Will merge on green.)

**Pull request merged**